### PR TITLE
freeze support

### PR DESCRIPTION
**Pull request by pranavk**
This PR adds support for synthesizing freeze instruction.

### CODE REVIEW

**Comment by pranavk** [comment]
I will start working on it once Freeze support in LLVM settles down. There are still some outstanding patches there.

**Comment by pranavk** [comment]
The CI will always fails because docker contains previous LLVM 9 version. Everything seems to be okay except one test `Solver/extract.ll` which is currently failing after I updated to LLVM 10 version in this PR. I am taking a look at it.

In the meantime, you can go ahead with the review.

**Comment by pranavk** [comment]
@zhengyangl I spent some time investigating why Solver/extract.ll is failing with LLVM 10 and it turns out this is due to LLVM 10 harvesting `range` metadata for one of the variable.

The test is:
```
%struct = type { i32, [4 x i8] }

define i32 @test1() {
  %A = extractvalue %struct { i32 2, [4 x i8] c"foo\00" }, 0
    ret i32 %A
    ; CHECK-LABEL: @test1(
    ; CHECK: ret i32 2
}
```
The difference in the code starts at this point (https://github.com/google/souper/blob/master/lib/Tool/CandidateMapUtils.cpp#L213). Cand.Mapping.LHS at this point for LLVM 9 was : 
`%0:i32 = var `
but with LLVM 10, it is:
`%0:i32 = var (range=[2,3))`. The new metadata is making L213 give us a non-null RHS while we were getting null RHS before. This is making the test to fail. The code goes deep into constant synthesis and diverges even further. You may have good insight on this one. Can you see where things may have broken or suggest where should I look?

**Comment by regehr** [comment]
good job tracking this down.
probably best to just replace this with a different file that tests the same thing, but doesn't allow LLVM to do the constantrange inference

**Comment by pranavk** [comment]
I am not sure this test is even necessary at all. Functionality of `extractvalue` is tested at several other places in the test suite as well.

**Comment by regehr** [comment]
fine to remove if zhengyang agrees this is OK

**Comment by zhengyangl** [comment]
Weird, I need to have a look before accepting the removal.

**Comment by zhengyangl** [comment]
TL DR: the test case extract.ll should be 
```
; REQUIRES: solver

; RUN: %llvm-as -o %t %s
; RUN: %souper %solver -check %t

target datalayout = "e-m:e-i64:64-f80:128-n8:16:32:64-S128"

%struct = type { i32, [4 x i8] }

define i32 @test1() {
  %A = extractvalue %struct { i32 2, [4 x i8] c"foo\00" }, 0, !expected !0
  ret i32 %A
}

!0 = !{i32 2}
```

**Comment by zhengyangl** [comment]
The original test case is problematic. It is designed to test if Souper can transform the extractvalue into a constant. It tries to use the FileCheck format for the text comparison, but on the header of the test case, there is no call to FileCheck in the `RUN` line, so it never works the way it is designed to do.

**Comment by zhengyangl** [comment]
The `souper` command with `-check` checks if the the synthesized const is equal to the `expected` metadata, it only fails the test case when they are mismatch or the `expected` metadata for the instruction is missing. If there is no synthesized const for the LHS, the test case passes. Before bumping to new LLVM, the `souper -check` succeed because Souper cannot synthesized the constant 2 from the extractvalue (the value is harvested as var), so the test case passes. But after bumping, LLVM assigns range DFA fact to the extractvalue, and Souper is now able to synthesize the constant with the range DFA fact. Unfortunately, there is no `expected` metadata assigned, thus the test cases fails.

**Comment by zhengyangl** [comment]
The fix is easy: add expected metadata and remove the unnecessary CHECK lines.

**Comment by pranavk** [comment]
Great. Thank you for the investigation. I will fix this.

**Comment by pranavk** [comment]
@regehr This PR is now ready to review. I have tested both release and debug builds and all tests passes.

File: lib/Extractor/KLEEBuilder.cpp [hunk 1]
```diff
@@ -18,6 +18,7 @@
 #include "souper/Extractor/ExprBuilder.h"
 #include "llvm/ADT/SmallVector.h"
 #include "llvm/Analysis/LoopInfo.h"
+#include "llvm/Support/CommandLine.h"
```
Comment (regehr) [inline lib/Extractor/KLEEBuilder.cpp]: what's this for?

File: lib/Extractor/KLEEBuilder.cpp [hunk 2]
```diff
@@ -141,6 +142,8 @@ class KLEEBuilder : public ExprBuilder {
       }
       return E;
     }
+    case Inst::Freeze:
+      return get(Ops[0]);
```
Comment (regehr) [inline lib/Extractor/KLEEBuilder.cpp]: is this just ignoring freeze in kleebuilder?

File: lib/Infer/Interpreter.cpp [hunk 3]
```diff
@@ -426,6 +426,13 @@ namespace souper {
       else
         return {ARG0.getHiBits(1).trunc(1)};
     }
+    case Inst::Freeze: {
+      if (Args[0].K == EvalValue::ValueKind::Undef || Args[0].K == EvalValue::ValueKind::Poison) {
+        return {llvm::APInt(Args[0].BitWidth,
+                std::rand() % llvm::APInt::getAllOnesValue(Args[0].BitWidth).getLimitedValue())};
```
Comment (regehr) [inline lib/Infer/Interpreter.cpp]: is this the right thing to do? it seems like this will make testing non-deterministic, etc.

File: test/Codegen/argument-order-XFAIL.opt [hunk 4]
```diff
@@ -10,7 +10,7 @@ result %4
 ; CHECK: ; ModuleID = 'souper.ll'
 ; CHECK-NEXT: source_filename = "souper.ll"
 
-; CHECK: define i8 @fun(i16, i4) {
+; CHECK: define i8 @fun(i16 %0, i4 %1) {
```
Comment (regehr) [inline test/Codegen/argument-order-XFAIL.opt]: I wasn't expecting you to need to do this, can you explain?

**Comment by regehr** [comment]
ok thanks. looks good overall, left a few comments.
can you please add a few more test cases for freeze though?

File: test/Codegen/argument-order-XFAIL.opt (hunk 4 above)
Comment (regehr) [reply test/Codegen/argument-order-XFAIL.opt]: ah nevermind, I can't believe they flipped this behavior again

File: lib/Extractor/KLEEBuilder.cpp (hunk 1 above)
Comment (pranavk) [reply lib/Extractor/KLEEBuilder.cpp]: The file uses cl:: namespace without including this header file. It worked with LLVM 9 probably because CommandLine.h was indirectly added by some other header file we are including here but this doesn't work with LLVM 10. Better to be explicit, as they say.

File: lib/Infer/Interpreter.cpp (hunk 3 above)
Comment (pranavk) [reply lib/Infer/Interpreter.cpp]: Yes, this will make testing hard to do. But I don't think it's right idea either to make `freeze` always come up with same result?

File: lib/Extractor/KLEEBuilder.cpp (hunk 2 above)
Comment (pranavk) [reply lib/Extractor/KLEEBuilder.cpp]: Sort of. I cannot put a explicit assert here to crash (as you suggested) as it results in several test failures because freeze is also synthesized now. So, it does need some sort of klee support for those RHSes to interpret successfully.

File: lib/Infer/Interpreter.cpp (hunk 3 above)
Comment (regehr) [reply lib/Infer/Interpreter.cpp]: Well, this needs to be thought about carefully, since it could have an impact on soundness. Talk to @manasij7479 about it please.

Comment (manasij7479) [reply lib/Infer/Interpreter.cpp]: Not really sure what to do here, other than: https://xkcd.com/221/  
Maybe a 'freeze fuzzing' mode which runs the tests with freeze producing arbitrary results?

Comment (regehr) [reply lib/Infer/Interpreter.cpp]: what all are we using the interpreter for? the solution we choose has to be appropriate for that.

Comment (manasij7479) [reply lib/Infer/Interpreter.cpp]: We interpret both sides to see if they can produce a different result.
If all freezes produce the same result, this technique will not rule out some invalid transformations, which will then be sent off to the solver.
Similarly for pruning, we interpret the LHS and abstract-interpret the RHS.
It seems to me that producing the same value will (rarely, if at all) inhibit pruning, not introduce unsoundness.

**Comment by pranavk** [comment]
Okay, finally was able to synthesize freeze. I added the test case above. There were some minor modifications I made to enumerative algorithm as well to prevent it from generating freezes of freezes.

File: include/souper/Infer/Interpreter.h [hunk 5]
```diff
@@ -48,14 +51,20 @@ struct EvalValue {
     return Value;
   }
 
-  static EvalValue poison() {
+  static EvalValue poison(unsigned Width) {
+    if (Width == 0)
```
Comment (regehr) [inline include/souper/Infer/Interpreter.h]: since there is not supposed to be a situation where this happens, please make it an assertion, not a regular error check (error checks are for things that can happen)

File: include/souper/Infer/Interpreter.h [hunk 6]
```diff
@@ -48,14 +51,20 @@ struct EvalValue {
...
+      llvm::report_fatal_error("poison of zero width!");
     EvalValue Result;
     Result.K = ValueKind::Poison;
+    Result.BitWidth = Width;
     return Result;
   }
-  static EvalValue undef() {
+  static EvalValue undef(unsigned Width) {
+    if (Width == 0)
```
Comment (regehr) [inline include/souper/Infer/Interpreter.h]: same

File: lib/Extractor/KLEEBuilder.cpp (hunk 2 above)
Comment (regehr) [reply lib/Extractor/KLEEBuilder.cpp]: we don't want silent unsound behavior, so something besides this needs to be done here

Comment (regehr) [reply lib/Extractor/KLEEBuilder.cpp]: you could add a flag to not synthesize freeze, and use it in the tests that crash

File: test/Infer/Interpreter/freeze.opt [hunk 7]
```diff
@@ -0,0 +1,12 @@
+; RUN: %souper-interpret %solver -input-values=%0=1 %s > %t 2>&1
+; RUN: %FileCheck -check-prefix=TEST1 %s < %t
+; TEST1: Value: 126
+
+; RUN: %souper-interpret %solver -input-values=%0=5 %s > %t 2>&1
+; RUN: %FileCheck -check-prefix=TEST2 %s < %t
+; TEST2-NOT: Value: Poison
```
Comment (regehr) [inline test/Infer/Interpreter/freeze.opt]: I do not understand this test case: a freeze instruction can never generate poison

**Comment by regehr** [comment]
looks good except for noted items, and also please switch to the LLVM 10 branch

File: include/souper/Infer/Interpreter.h (hunk 5 above)
Comment (pranavk) [reply include/souper/Infer/Interpreter.h]: this is to make sure that even if we do a release build without assertions, it doesn't go unnoticed. I vaguely remember some time back when we changed the asserts to this "fatal error" to make sure we don't skip asserts under any circumstances.

File: test/Infer/Interpreter/freeze.opt (hunk 7 above)
Comment (pranavk) [reply test/Infer/Interpreter/freeze.opt]: This is what this test is asserting. Note the "-NOT" after TEST2. "-NOT" means that it never generates poison.

File: include/souper/Infer/Interpreter.h (hunk 5 above)
Comment (regehr) [reply include/souper/Infer/Interpreter.h]: souper has a lot of assertions in it, we never got rid of them

Comment (regehr) [reply include/souper/Infer/Interpreter.h]: if you're simply guarding against internal inconsistency, assertion is usualyl the way to go, and it is what you should be using here

File: test/Infer/Interpreter/freeze.opt (hunk 7 above)
Comment (regehr) [reply test/Infer/Interpreter/freeze.opt]: ah, got it

File: include/souper/Infer/Interpreter.h (hunk 5 above)
Comment (pranavk) [reply include/souper/Infer/Interpreter.h]: ok. i will change it to assert.

**Comment by regehr** [comment]
thanks!
